sdk/core/System.ClientModel/src/Convenience/ClientCache.cs

```diff
@@ -30,22 +30,30 @@ public ClientCache(int maxSize = 100)
     /// Updates the last-used timestamp on every hit.
     /// </summary>
     /// <typeparam name="T">The type of the client.</typeparam>
-    /// <param name="clientId">An equality-comparable key representing the client configuration.</param>
+    /// <param name="clientId">A key representing the client configuration.</param>
     /// <param name="createClient">A factory function to create the client if not cached.</param>
     /// <returns>The cached or newly created client instance.</returns>
-    public T GetClient<T>(IEquatable<object> clientId, Func<T> createClient) where T : class
+    public T GetClient<T>(object clientId, Func<T> createClient) where T : class
     {
-        // If the client exists, update its timestamp.
-        if (_clients.TryGetValue(clientId, out var cached))
+        _lock.EnterReadLock();
+        try
         {
-            cached.LastUsed = Stopwatch.GetTimestamp();
-
-            if (cached.Client is T typedClient)
+            // If the client exists, update its timestamp.
+            if (_clients.TryGetValue(clientId, out var cached))
             {
-                return typedClient;
-            }
+                cached.LastUsed = Stopwatch.GetTimestamp();
+
+                if (cached.Client is T typedClient)
+                {
+                    return typedClient;
+                }
 
-            throw new InvalidOperationException($"The clientId is associated with a client of type '{cached.Client.GetType()}', not '{typeof(T)}'.");
+                throw new InvalidOperationException($"The clientId is associated with a client of type '{cached.Client.GetType()}', not '{typeof(T)}'.");
+            }
+        }
+        finally
+        {
+            _lock.ExitReadLock();
```

Comment:
since you exiting the lock here, the following lines that create new client might execute more than once for the same clientId (they threads will just wait to do it sequentially). It's probably not a big issue, but if we want to avoid it, just check if the ID exists in the dictionary once you enter the write lock

Reply:
Thanks for pointing this out! I've updated the impl to recheck the dictionary under the write lock.